server/config/index.js

```diff
@@ -7,18 +9,46 @@ const data = require( './parser' )( configPath, {
 } );
 
 /**
- * Return config `key`.
- * Throws an error if the requested `key` is not set in the config file.
+ * Returns configuration value for given key
+ *
+ * If the requested key isn't defined in the configuration
+ * data then this will report the failure with either an
+ * error or a console warning.
+ *
+ * When in the 'development' NODE_ENV it will raise an error
+ * to crash execution early. However, because many modules
+ * call this function in the module-global scope a failure
+ * here can not only crash that module but also entire
+ * subsystems within Calypso. Therefore if the NODE_ENV is
+ * not 'development' we will return `undefined` and log a
+ * message to console instead of halting the execution thread.
  *
- * @param {String} key The key of the config entry.
- * @return {Mixed} Value of config or error if not found.
- * @api public
+ * The config files are loaded in sequence: _shared.json, {env}.json, {env}.local.json
+ * @see server/config/parser.js
+ *
+ * @throws {ReferenceError} when key not defined in the config (NODE_ENV=development only)
+ * @param {String} key name of the property defined in the config files
+ * @returns {*} value of property named by the key
  */
 function config( key ) {
 	if ( key in data ) {
 		return data[ key ];
 	}
-	throw new Error( 'config key `' + key + '` does not exist' );
+
+	const errorMessage = (
+		`Could not find config value for key '${ key }'\n` +
+		`Please make sure that if you need it then it has a value assigned in 'config/_shared.json'`
+	);
+
+	if ( 'development' === process.env.NODE_ENV ) {
+		throw new ReferenceError( errorMessage );
+	}
+
+	// display console error only in a browser
+	// (not in tests, for example)
+	if ( 'undefined' !== typeof window ) {
+		console.error( errorMessage );
```

Comment:
>it's my belief right now that we want to make an annoying console message so that any of us can quickly identify the failures in production but we want to prevent those failures from cascading outward thus removing the error

Makes sense 👍